### ISSUE TITLE
Graphcache performance issues with big lists

### ISSUE THREAD

**Issue by Jylth** [FormidableLabs/urql#1977]
Hello,

We seem to have been hitting a wall on some use cases in our mobile app. Graphcache seems to struggle quite a lot on mid-tier devices with lists above 300-500 items. The UI is getting completely blocked by the simplest mutation on a single item.

urql version & exchanges:

urql : 2.0.4
urql/exchange-graphcache: 4.3.3

**Steps to reproduce**

I set up a sandbox which shows how the cache is getting slower to update over time when we are adding more items to the list. After loading all the 10000 items, a mutation on a single item can take up to several seconds: try increasing the quantity of an item through the form input. The list is only displaying 2 items, so we can clearly see this is not a React rendering issue.

I used a faster implementation of the simplePagination to handle pagination, but even without any kind of pagination, the issue is still there.

Chrome seems to be better at handling this. Firefox not so much. The real issue arises on a mobile environment with limited resources. A list with only 500 items is a complete hassle to fetch & update on React Native, even on newer devices. To emulate this, you should run the sandbox on Google Chrome with a 4x CPU throttle.

[sandbox link](https://codesandbox.io/s/fervent-worker-3p86l?file=/src/index.js)

**Expected behavior**

The performance of updating a single item in a list should not depend on the size of the list. I would expect a list of any size to update at the same rate. The normalized cache should enable this type of performance.

Blocking the JS thread for more than a hundred ms is too much if we want to build responsive & fast apps, expensive background processes should pass priority, through [setImmediate](https://developer.mozilla.org/en-US/docs/Web/API/Window/setImmediate ), to avoid making the app unresponsive.

**Actual behavior**

A simple mutation on a single item is blocking the whole JS thread, making the app unresponsive for a couple seconds.

Thanks for your hard work, besides this issue, the rest of the framework is a pleasure to use ! Hope we can fix this fast :). 

I started to profile where the bottlenecks are located,  will keep this post updated.

**Comment by kitten** [comment]
Ok, I haven't yet looked at the entire thing, but this overall looks extremely odd given that there's no optimistic layers, no special tasks, and a couple of queries, however, it'll take me a bit to untangle this and extract the relevant parts. Can you hint at what is taking longer than expected though? For reference, we have benchmarks executing similar conditions in the 10,000 range, i.e. where we actually are writing/querying 10,000 items at once (without taking shortcuts), so this seems odd

Can you leave some more notes here please on the performance characteristics you're seeing in different environments? That'd scope this down for us, since you already hace the numbers.

> The performance of updating a single item in a list should not depend on the size of the list.

That's very true, however, each time a single item updates in a list, the list needs to be requeried. This shouldn't take longer however than a fraction of a second however even for _huge_ lists, i.e. 10,000 items, so if your query doesn't actually include that many items, it'll be extremely fast — in fact, in Graphcache that's an order of magnitude faster than [insert other caching client here] 😆 

> Blocking the JS thread for more than a hundred ms is too much if we want to build responsive & fast apps, expensive background processes should pass priority, through setImmediate, to avoid making the app unresponsive.

It's unfortunately simply _not_ that simple 😅 And the React team, for instance, is a team that knows this well. Any given task shouldn't take long (which is why we optimise our tasks to be as fast as they can be), however, data tasks are special. They cannot be broken up (each query needs to be executed as is, consistently) and dependent data cannot be separated (if data changes, we don't want to "fracture" or "tear" the UI by showing different data in different places).

As a data library, (especially one that caters for different cases that aren't just React) we also cannot defer many tasks. When you execute a cached query, you'll need the result immediately. When a mutation resolves, you'll need to store it immediately. Given the other preconditions there isn't a simple defer that can be thrown in. For comparison though, our storage and GC logic is deferred however. (although not chunked, since it's mostly an even smaller task than querying/writing)

I'll simply have to see what's going on here, and once we untangle that we can hopefully find out why the time here is clearly slowing with more items, although we aren't asking for a huge list (if that's actually the case 😆 )

**Comment by kitten** [comment]
Ok, _very_ quick update, that was actually a quick one; I'm seeing that the sandbox renders:

> todos total = 10000

And that corresponds to the amount of items that we're querying. 10,000 items are going to take _some time_, and mutations that edit one (or add one) will lead to the entire query having to rerun. That's simply because you can't artifically skip over the querying process — even if we did, that wouldn't exactly be possible once a resolver is in play, because any part of the query may change 😅 For a thought experiment, if we kept around a mapping of where certain entities are in each query, we'd build up a tremendously big map of "jump points" into the data and into the GraphQL document, which is a huge amount of complexity for something that may not speed up exponentially.

In short though; what we should see is sub-100ms runs in optimal conditions for 10,000 items, if I'm remembering correctly. That's a huge amount of data and it does need to be updated at some point 😆 There's not much to "fix" here in that regards because of the fact that it's such a large amount. It's possible we could skip over some updates in the future, but it'd scale this time down by a smaller factor than an order of magnitude;
meaning, if you'd then query 100,000 items then it'd be slow again. Now, we may say "100,000? That's ridiculous" 😛 , however, that's already the case at 10,000. What can be done however is that you could use a different method of pagination, which is the one we'd usually recommend in more mature apps.

The method we'd lean to at such a high amount of items is separating pages of queries (rather than using a resolver) into separate queries altogether. https://github.com/FormidableLabs/urql/blob/3e8929ad56d2f162091c87aa5611818417dbd8fd/examples/with-pagination/src/PaginatedNpmSearch.jsx#L22-L27
This changes a lot of things, because we aren't artifically "forcing" Graphcache to combine all pages into a single huge query.

**Comment by Jylth** [comment]
I see ! Thanks once again for the fast response, you should definitely add this use case in the documentation. 
The graphcache can indeed handle a lot of simultaneous smaller queries, but struggles with those single, fat ones.

Happy to see the issue was more on how we used the tool :).

This issue should be resolved

**Comment by kitten** [comment]
Alright, we'll still come back to revisit this, but what we're thinking is, if you have a given query that's "huge" (like, basically, 100–1000x the cost of a regular one) and that one has a previous response, then we could skip some work given that only small parts of it may update. However, the problems with that are two-fold:

- We'd still have to run it at some point (and if the result is huge its first run must take long)
- We may run into cases where many updates still slow it down
- To skip over re-querying the entire huge query, we'd have to keep a complex data structure around that nestedly tracks data dependencies (we aren't 100% sure if that's even easily possible given N-M relations via resolvers and such)

So, while there are some optimisations in theory, we'd consider this risky in terms of what it may do to the codebase's complexity and hence bundlesize. We'll have to investigate whether there are some easier wins first, now that we do track the previously queried data 🙌

**Comment by kitten** [comment]
@Jylth I haven't experimented with this in a while, or with react-virtual etc which are newer approaches to virtualisation, but in theory any approach that supports dynamically sized items shouldn't only be able to deal with individual items but could also be reutilised to just deal with batches on their own. This is similar to how a lot of virtualisation libraries already work. A lot of them will define an amount of items that render together, which means that you can virtualise entire batches as a whole, which can then correspond to your pages.

**Comment by Jylth** [comment]
You're completely right ! We tried to fit this new way of using the data to our existing lists without taking the time to take a step back ^^. Thanks again for the quick & very effective response !
